source/nodejs/adaptivecards/src/card-elements.ts

```diff
@@ -26,19 +26,101 @@ function generateUniqueId(): string {
+    let result: Action = null;
 
-    var result = AdaptiveCard.actionTypeRegistry.createInstance(actionType);
+    if (json && typeof json === "object") {
+        let tryToFallback = false;
+        let actionType = json["type"];
 
-    if (!result) {
-        raiseParseError(
-            {
-                error: Enums.ValidationError.UnknownActionType,
-                message: "Unknown action type: " + actionType
-            },
-            errors
-        );
+        result = AdaptiveCard.actionTypeRegistry.createInstance(actionType);
+
+        if (!result) {
+            tryToFallback = true;
+
+            raiseParseError(
+                {
+                    error: Enums.ValidationError.UnknownActionType,
+                    message: "Unknown action type: " + actionType + ". Attempting to fall back."
+                },
+                errors
+            );
+        }
+        else {
+            result.setParent(parent);
+            result.parse(json);
+
+            tryToFallback = result.shouldFallback();
+        }
+
+        if (tryToFallback) {
+            let fallback = json["fallback"];
+
+            if (!fallback) {
+                parent.setShouldFallback(true);
+            }
+            if (typeof fallback === "string" && fallback.toLowerCase() === "drop") {
+                result = null;
+            }
+            else if (typeof fallback === "object") {
+                result = createActionInstance(
+                    parent,
+                    fallback,
+                    errors);
+            }
+        }
+    }
+
+    return result;
+}
+
+export function createElementInstance(
+    parent: CardElement,
+    json: any,
+    errors: Array<IValidationError>): CardElement {
+    let result: CardElement = null;
+
+    if (json && typeof json === "object") {
+        let tryToFallback = false;
+
+        result = AdaptiveCard.elementTypeRegistry.createInstance(json["type"]);
+
+        if (!result) {
+            tryToFallback = true;
+
+            raiseParseError(
+                {
+                    error: Enums.ValidationError.UnknownElementType,
+                    message: "Unknown element type: " + json["type"] + ". Attempting to fall back."
+                },
+                errors
+            );
+        }
+        else {
+            result.parse(json, errors);
+
+            tryToFallback = result.shouldFallback();
+        }
+
+        if (tryToFallback) {
+            let fallback = json["fallback"];
+
+            if (!fallback) {
+                parent.setShouldFallback(true);
+            }
+            if (typeof fallback === "string" && fallback.toLowerCase() === "drop") {
+                result = null;
+            }
+            else if (typeof fallback === "object") {
+                result = createElementInstance(
+                    parent,
+                    fallback,
+                    errors);
+            }
+        }
```

Comment:
a lot of this logic is duplicated from above. helper function?

Reply:
Done